### PR TITLE
Parse/save dungeon data files

### PR DESCRIPTION
**Pull request by StrikerX3**
These changes allow the editor to access and modify `dungeon_data_info.bin`, `dungeon_extra.bin` and `dungeon_balance.bin/ent` files.

I've also added a script demonstrating how to interpret the data:
- Basic dungeon information such as number of floors, whether teammates/items are allowed, whether the dungeon resets the level, permanent Scanning/Radar status, etc.
- Events and other data for each floor of the dungeons
- Wild Pokémon spawns per floor, their stats, spawn rates and Strong Foe status
- Mystery House and fainted Pokémon listings per dungeon

I've also included a GYU0 decompressor (written by AntyMew) and my own implementation of a compressor for the format which implements all commands. However, the "copy from previously written data" command is disabled due to the slow algorithm currently in use.

### CODE REVIEW

File: SkyEditor.RomEditor.Rtdx/Domain/Common/Structures/Gyu0.cs [hunk 1]
```diff
@@ -0,0 +1,339 @@
...
+                    continue;
+                }
+
+                var op = (Opcode)((cur >> 5) - 4);
+                var arg = cur & 0x1F;
+
+                switch (op)
+                {
+                    case Opcode.Copy:
+                        {
```
Comment (evandixon) [inline SkyEditor.RomEditor.Rtdx/Domain/Common/Structures/Gyu0.cs:53]: Curly braces aren't required in switch statements. C# has extra rules preventing you from shooting yourself in the foot, but switch statements are just glorified goto's.

Comment (StrikerX3) [reply SkyEditor.RomEditor.Rtdx/Domain/Common/Structures/Gyu0.cs:53]: They are necessary in this case due to use of variables within case statements.

File: SkyEditor.RomEditor.Rtdx/Domain/Rtdx/Models/DungeonCollection.cs [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+            {
+                dungeons.Add(new DungeonModel(commonStrings, dungeon.Value)
+                {
+                    Id = dungeon.Key,
+                    Extra = dungeonExtra.Entries.GetValueOrDefault(dungeon.Key),
+                    Balance = dungeonBalance.Entries[dungeon.Value.DungeonBalanceIndex]
+                });
+            }
+            dungeons.Sort((d1, d2) => d1.Data.SortKey.CompareTo(d2.Data.SortKey));
+            return dungeons.ToArray();
```
Comment (evandixon) [inline SkyEditor.RomEditor.Rtdx/Domain/Rtdx/Models/DungeonCollection.cs:51]: I personally prefer `dungeons.OrderBy(d => d.Data.SortKey).ToArray()`, but there's no need to change it now.

File: SkyEditor.RomEditor.Rtdx/Domain/Rtdx/Structures/DungeonBalance.cs [hunk 3]
```diff
@@ -0,0 +1,692 @@
...
+                    bin.Position = bin.Length;
+                }
+                entryPointers.Add((int)bin.Position);
+            }
+
+            // Build the .ent file data
+            var ent = new byte[entryPointers.Count * sizeof(int)];
+            for (int i = 0; i < entryPointers.Count; i++)
+            {
+                BitConverter.GetBytes(entryPointers[i]).CopyTo(ent, i * sizeof(int));
```
Comment (evandixon) [inline SkyEditor.RomEditor.Rtdx/Domain/Rtdx/Structures/DungeonBalance.cs:74]: Alternatively,
```
BinaryPrimitives.WriteInt32LittleEndian(ent.AsSpan().Slice(i * sizeof(int)), entryPointers[i]);
```
This reduces the allocation of an array on the heap, although the performance gain will be negligible.

File: SkyEditor.RomEditor.Rtdx/Domain/Rtdx/Structures/FixedPokemon.cs [hunk 4]
```diff
@@ -60,12 +60,14 @@ public FixedPokemonEntry(int index, byte[] data, int offset)
             Move2 = (WazaIndex)BitConverter.ToInt16(data, offset + 0xA);
             Move3 = (WazaIndex)BitConverter.ToInt16(data, offset + 0xC);
             Move4 = (WazaIndex)BitConverter.ToInt16(data, offset + 0xE);
-            Level = data[0x16];
```
Comment (evandixon) [inline SkyEditor.RomEditor.Rtdx/Domain/Rtdx/Structures/FixedPokemon.cs:63]: I should have known better than to just assume this was the same as PSMD.

File: SkyEditor.RomEditor.Rtdx/Domain/Rtdx/Constants/DungeonIndex.cs [hunk 5]
```diff
@@ -0,0 +1,152 @@
+﻿using SkyEditor.RomEditor.Infrastructure.Automation.CSharp;
+using SkyEditor.RomEditor.Infrastructure.Automation.Lua;
+using SkyEditor.RomEditor.Infrastructure;
+using System;
+using System.Collections.Generic;
+
+namespace SkyEditor.RomEditor.Domain.Rtdx.Constants
+{
+    public class DungeonIndexLuaExpressionGenerator : ILuaExpressionGenerator
```
Comment (evandixon) [inline SkyEditor.RomEditor.Rtdx/Domain/Rtdx/Constants/DungeonIndex.cs:9]: My code isn't the best example of consistency, but it'd be best if either both code generators were in this file or if they were both in the same file as the generators, and not a mix of the two.

But I'm definitely guilty of this too.

**Review by evandixon — APPROVED** [review]

**Pull request merged**